modules/security/src/main/java/io/oasp/module/security/common/base/accesscontrol/AbstractAccessControlProvider.java

```diff
@@ -63,17 +67,19 @@ protected void initialize(AccessControlSchema config) {
    * graph}.
    *
    * @param group is the {@link AccessControlGroup} to check.
-   * @param visitedGroups is where the {@link AccessControlGroup} are collected to detect cycles.
+   * @param nodeCycle the {@link NodeCycle} used to detect cycles.
    */
-  protected void checkForCyclicDependencies(AccessControlGroup group, Set<AccessControlGroup> visitedGroups) {
+  protected void checkForCyclicDependencies(AccessControlGroup group, NodeCycle<AccessControlGroup> nodeCycle) {
 
-    boolean added = visitedGroups.add(group);
-    if (!added) {
-      // mmm NodeCycleException would be very helpful here...
-      throw new IllegalStateException("Cyclic inheritance of access control groups detected for " + group);
-    }
     for (AccessControlGroup inheritedGroup : group.getInherits()) {
-      checkForCyclicDependencies(inheritedGroup, visitedGroups);
+      List<AccessControlGroup> inverseCycle = nodeCycle.getInverseCycle();
```

Comment:
It is not clear for me why you need this NodeCycle stuff. The class claimed, that it is a datastructure to detect cycles. However, you are just simulating a stack here, It is somehow confusing to work with a "cycle" here, which does not represent a cycle at all. In addition to that, inverseCycle is even more strange especially as the javaDoc of this function does not state what the return value of this method will be if there is no cycle. 
Is it simply the exception to be reused here?

Maybe `visitedGroups` is not the correct name anymore, but `nodeCycle` neither. It does not has anything in common with the value of the parameter. More suitable would be visitedStack or something like this.